exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicSchema.java

```diff
@@ -32,23 +33,34 @@
  */
 public class DynamicSchema extends SimpleCalciteSchema implements AutoCloseable {
 
-  public DynamicSchema(CalciteSchema parent, Schema schema, String name) {
+  private UserSession session;
+
+  public DynamicSchema(CalciteSchema parent, Schema schema, String name, UserSession session) {
     super(parent, schema, name);
+    this.session = session;
   }
 
   @Override
   protected CalciteSchema getImplicitSubSchema(String schemaName,
                                                boolean caseSensitive) {
     Schema s = schema.getSubSchema(schemaName);
     if (s != null) {
-      return new DynamicSchema(this, s, schemaName);
+      return new DynamicSchema(this, s, schemaName, session);
     }
     return getSubSchemaMap().get(schemaName);
   }
 
+  public UserSession getSession() {
+    return session;
+  }
+
+  public void setSession(UserSession session) {
```

Comment:
I think we shouldn't set a session here also, just use the one from the constructor.

Reply:
Fixed.  `UserSession` is now passed only in constructors.

Reply:
Fixed... the `UserSession` is only updated in the constructor.  There is now a `clone()` method which also updates the  UserSession and an additional constructor.